### PR TITLE
HAAR-819: Move 'api/groups/{group}' from 'hmpps-auth' to external use…

### PR DESCRIPTION
**Pull request by ravmoj**
Move 'api/groups/{group}' from 'hmpps-auth' to external user api

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/service/GroupsService.kt [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+    Hibernate.initialize(requestedGroup.assignableRoles)
+    Hibernate.initialize(requestedGroup.children)
+    maintainUserCheck.ensureMaintainerGroupRelationship(groupCode)
+    return requestedGroup
+  }
+
+    /* @Throws(ChildGroupNotFoundException::class)
+     fun getChildGroupDetail(
+       groupCode: String,
```
Comment (karenmillermoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/service/GroupsService.kt:30]: you have tests that still need to be uncommented

File: src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/jpa/repository/UserRepository.kt [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+import org.springframework.data.jpa.repository.JpaSpecificationExecutor
+import org.springframework.data.jpa.repository.Query
+import org.springframework.data.repository.CrudRepository
+import uk.gov.justice.digital.hmpps.externalusersapi.model.User
+import uk.gov.justice.digital.hmpps.externalusersapi.security.AuthSource
+import java.util.Optional
+import java.util.UUID
+
+interface UserRepository : CrudRepository<User, UUID>, JpaSpecificationExecutor<User> {
+  // fun findByUsername(username: String?): Optional<User>
```
Comment (karenmillermoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/jpa/repository/UserRepository.kt:12]: should be removed if no longer used

File: src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/service/GroupsService.kt (hunk 1 above)
Comment (ravmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/service/GroupsService.kt]: Removed unrelated (commented) code

File: src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/jpa/repository/UserRepository.kt (hunk 2 above)
Comment (ravmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/jpa/repository/UserRepository.kt]: Removed commented code

File: src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/jpa/repository/UserRepository.kt [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+import org.springframework.data.jpa.repository.JpaSpecificationExecutor
+import org.springframework.data.jpa.repository.Query
+import org.springframework.data.repository.CrudRepository
+import uk.gov.justice.digital.hmpps.externalusersapi.model.User
+import uk.gov.justice.digital.hmpps.externalusersapi.security.AuthSource
+import java.util.Optional
+import java.util.UUID
+
+interface UserRepository : CrudRepository<User, UUID>, JpaSpecificationExecutor<User> {
+  @Query("select distinct u from User u left join fetch u.authorities where u.username = :username and u.source = :source")
```
Comment (AlTurner-MOJ) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/jpa/repository/UserRepository.kt:12]: In our context AuthSource will always be external. We should be able to remove the AuthSource enum and just hard code the filter in the query for now?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/security/AuthSource.kt [hunk 4]
```diff
@@ -0,0 +1,25 @@
+package uk.gov.justice.digital.hmpps.externalusersapi.security
+
+import com.fasterxml.jackson.annotation.JsonValue
+
+enum class AuthSource(val description: String) {
```
Comment (AlTurner-MOJ) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/security/AuthSource.kt:5]: As mentioned above this enum feels unnecessary within external-users-api.

**Review by AlTurner-MOJ — CHANGES_REQUESTED** [review]

File: src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/security/AuthSource.kt (hunk 4 above)
Comment (ravmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/security/AuthSource.kt:5]: Will take this clean up exercise after DEVs discussion tomorrow

File: src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/jpa/repository/UserRepository.kt (hunk 3 above)
Comment (ravmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/jpa/repository/UserRepository.kt:12]: The actual method invoking has Authsource hardcoded as 
'findByUsernameAndSource(username, AuthSource.auth)'

File: src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/config/HmppsExternalUsersApiExceptionHandler.kt [hunk 5]
```diff
@@ -103,6 +105,47 @@ class HmppsExternalUsersApiExceptionHandler {
...
+      .status(HttpStatus.FORBIDDEN)
+      .body(
+        ErrorResponse(
+          status = HttpStatus.FORBIDDEN,
+          userMessage = "Auth user group relationship exception: ${e.message}",
+          developerMessage = e.message ?: "Error message not set"
+        )
+      )
+  }
+
```
[the remark this answers is not in the document]
Comment (ravmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/config/HmppsExternalUsersApiExceptionHandler.kt:134]: Done

File: src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/config/HmppsExternalUsersApiExceptionHandler.kt [hunk 6]
```diff
@@ -103,6 +105,47 @@ class HmppsExternalUsersApiExceptionHandler {
...
+      .status(NOT_FOUND)
+      .body(
+        ErrorResponse(
+          status = NOT_FOUND,
+          userMessage = "Group Not found: ${e.message}",
+          developerMessage = e.message ?: "Error message not set"
+        )
+      )
+  }
+  @ExceptionHandler(MaintainUserCheck.AuthUserGroupRelationshipException::class)
```
[the remark this answers is not in the document]
Comment (ravmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/config/HmppsExternalUsersApiExceptionHandler.kt:121]: Done

**Review by AlTurner-MOJ — APPROVED** [review]
lgtm

**Pull request merged**